File: javascript/708.js
/**
 * // Definition for a Node.
 * function Node(val,next) {
 *    this.val = val;
 *    this.next = next;
 * };
 */
/**
 * @param {Node} head
 * @param {number} insertVal
 * @return {Node}
 */

var insert = function(head, insertVal) {
    if(head == null) {
        let ret = new Node(insertVal, null);
        ret.next = ret;
        return ret;
    }
    let cur = head.next;
    while(cur.val <= cur.next.val && cur.next !== head) cur = cur.next;
    if(insertVal >= cur.val){ //bigger than any node val
        cur.next = new Node(insertVal, cur.next);
        return head;
    }
    // let biggest = cur;
    while(insertVal > cur.next.val) cur = cur.next;
    cur.next = new Node(insertVal, cur.next);
    return head;
};

// let n1 = new Node(3, null);
// let n2 = new Node(1, n1);
// let n3 = new Node(4, n2);
// n1.next = n3;
// insert(n1, 2);
// let cur = n2;
// console.log(n1.val);
// while(cur !== n1) {
//     console.log(cur.val);
//     cur = cur.next;
// }
// let n1 = insert(null, 1);
// console.log(n1.next.val);

File: C++/code/708.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;

    Node() {}

    Node(int _val) {
        val = _val;
        next = NULL;
    }

    Node(int _val, Node* _next) {
        val = _val;
        next = _next;
    }
};
*/

class Solution {
public:
    Node* insert(Node* head, int insertVal) {
        if (!head) {
            head = new Node(insertVal);
            head->next = head;
            return head;
        } 
        Node* lastNode = head;
        while (lastNode->next != head) {
            if (insertVal <= lastNode->next->val &&
               (lastNode->val <= insertVal || lastNode->val > lastNode->next->val)) break;
            if (lastNode->val > lastNode->next->val && insertVal >= lastNode->val) break;
            lastNode = lastNode->next;
        }
        Node* node = new Node(insertVal);
        node->next = lastNode->next;
        lastNode->next = node;
        return head;
    }
};


// 1 2 3 4 5